homeassistant/components/mobile_app/__init__.py

```diff
@@ -0,0 +1,396 @@
+        return Response(status=200)
+
+
+def supports_encryption():
+    """Test if we support encryption."""
+    try:
+        import nacl   # noqa pylint: disable=unused-import
+        return True
+    except OSError:
+        return False
+
+
+def device_for_webhook_id(hass, webhook_id):
+    """Return the device name for the webhook ID."""
+    for device_name, device in hass.data[DOMAIN].items():
+        if device.get(CONF_WEBHOOK_ID) == webhook_id:
+            return device
+    return None
+
+
+def safe_device(device: dict):
+    """Return a device without webhook_id or secret."""
+    return {
+        ATTR_APP_ID: device[ATTR_APP_ID],
+        ATTR_APP_NAME: device[ATTR_APP_NAME],
+        ATTR_APP_VERSION: device[ATTR_APP_VERSION],
+        ATTR_FINGERPRINT: device[ATTR_FINGERPRINT],
+        ATTR_DEVICE_NAME: device[ATTR_DEVICE_NAME],
+        ATTR_INTEGRATION_DATA: device[ATTR_INTEGRATION_DATA],
+        ATTR_MANUFACTURER: device[ATTR_MANUFACTURER],
+        ATTR_MODEL: device[ATTR_MODEL],
+        ATTR_OS_VERSION: device[ATTR_OS_VERSION],
+        ATTR_SUPPORTS_ENCRYPTION: device[ATTR_SUPPORTS_ENCRYPTION],
+    }
+
+
+async def async_setup(hass, config):
+    """Set up the mobile app component."""
+    conf = config.get(DOMAIN)
+
+    store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+    app_config = await store.async_load()
+    if app_config is None:
+        app_config = {}
+
+    hass.data[DOMAIN] = app_config
+
+    for key, device in app_config.items():
+        webhook.async_register(hass, DOMAIN, 'Mobile app',
+                               device[CONF_WEBHOOK_ID], handle_webhook)
+
+    if conf is not None:
+        hass.async_create_task(hass.config_entries.flow.async_init(
+            DOMAIN, context={'source': config_entries.SOURCE_IMPORT}))
+
+    hass.http.register_view(DevicesView(store))
+    hass.http.register_view(SingleDeviceView(store))
+
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up an mobile_app entry."""
+    return True
+
+
+class DevicesView(HomeAssistantView):
+    """A view that accepts device registration requests."""
+
+    url = '/api/mobile_app/devices'
+    name = 'api:mobile_app:register-device'
+
+    def __init__(self, store):
+        """Initialize the view."""
+        self._store = store
+
+    async def get(self, request):
+        """Handle a request to get all registered devices."""
+        all_devices = []
+
+        hass = request.app['hass']
+
+        for key, device in hass.data[DOMAIN].items():
+            all_devices.append(safe_device(device))
+
+        return self.json(all_devices)
+
+    @RequestDataValidator(REGISTER_DEVICE_SCHEMA)
+    async def post(self, request, data):
+        """Handle the POST request for device registration."""
+        hass = request.app['hass']
+
+        fingerprint = data[ATTR_FINGERPRINT]
+
+        if fingerprint in hass.data[DOMAIN]:
```

Comment:
I'm not sure we even need a fingerprint. An app should keep track of login info, if they forget that, they should just get a new registration.

Reply:
It simplifies things and I don't see any functionality loss

Reply:
Sorry, the reasoning around `fingerprint` was in one of the previously collapsed comments [here](https://github.com/home-assistant/home-assistant/pull/21475#discussion_r261497201)